### PR TITLE
[HttpFoundation] enable HTTP method overrides as early as possible with the HTTP cache

### LINKED ISSUE #40452: submitting Form wiht PATCH method (isSubmitted()) not working
**Symfony version(s) affected**: 5.2.3

**Description**

on a Form with a PATCH method, with the http_method_override options set to true, the form isn't marked as submitted.
The method property of the `Request` class is already set to POST and so it doesn't pass the first condition of the `handleRequest()` method.

**How to reproduce**

set the http_method_override to true and the forme method to PATCH.

**Possible Solution**

I didn't really look at it in details yet, but from the (dirty) debug I made, the method property of the `Request` class is already set to POST and so it doesn't pass the first condition of the `handleRequest()` method.

**Additional context**

`framework.yml`
```
framework:
  secret: '%env(APP_SECRET)%'
  #default_locale: en
  csrf_protection: ~
  http_method_override: true
```

`MyForm.php`
```
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->setMethod('PATCH');
        ...
```

`MyController`
```
 public function edit(Request $request)
    {
        $form = $this->createForm(MyForm::class, $class);
        $form->handleRequest($request);

        if ($form->isSubmitted()) { // Always false when using PATCH, works when using POST
```

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #40452
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I don't think we should do this on 4.4. On 4.4, it should be ppl's job to call the method early enough.

On 5.2, we should call the method when constructing the `http_cache` service.

**Comment by xabbuh** [comment]
I think you are right. This also fixes the cache key issue.

File: src/Symfony/Bundle/FrameworkBundle/HttpCache/HttpCache.php [hunk 1]
```diff
@@ -62,6 +63,15 @@ public function __construct(KernelInterface $kernel, $cache = null, SurrogateInt
         parent::__construct($kernel, $this->createStore(), $this->createSurrogate(), array_merge($this->options, $this->getOptions()));
     }
 
+    public function handle(Request $request, int $type = HttpKernelInterface::MASTER_REQUEST, bool $catch = true)
+    {
+        if ($this->kernel->getContainer()->getParameter('kernel.http_method_override')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/HttpCache/HttpCache.php:68]: `if (($c = $this->kernel->getContainer())->hasParameter('kernel.http_method_override') && $c->getParameter('kernel.http_method_override')) {`

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/HttpCache/HttpCache.php:68]: nvm, this is in fwb, where the parameter is always defined

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**